pkg/iobuf/bufferedreaderseeker.go

```diff
@@ -0,0 +1,198 @@
+	var (
+		seeker          io.Seeker
+		buffer          *bytes.Buffer
+		activeBuffering = true
+	)
+	if s, ok := r.(io.Seeker); ok {
+		seeker = s
+		activeBuffering = false
+	}
+
+	const mimeTypeBufferSize = 3072 // Approx buffer size for MIME type detection
+
+	if seeker == nil {
+		buffer = bytes.NewBuffer(make([]byte, 0, mimeTypeBufferSize))
+	}
+
+	return &BufferedReadSeeker{
+		reader:          r,
+		seeker:          seeker,
+		buffer:          buffer,
+		bytesRead:       0,
+		index:           0,
+		activeBuffering: activeBuffering,
+	}
+}
+
+// Read reads len(out) bytes from the reader starting at the current index.
+// It handles both seekable and non-seekable underlying readers efficiently.
+func (br *BufferedReadSeeker) Read(out []byte) (int, error) {
+	// For seekable readers, read directly from the underlying reader.
+	if br.seeker != nil {
+		n, err := br.reader.Read(out)
+		br.index += int64(n)
+		br.bytesRead = max(br.bytesRead, br.index)
+		return n, err
+	}
+
+	// For non-seekable readers, use buffered reading.
+	outLen := int64(len(out))
+	if outLen == 0 {
+		return 0, nil
+	}
+
+	// If the current read position (br.index) is within the buffer's valid data range,
+	// read from the buffer. This ensures previously read data (e.g., for mime type detection)
+	// is included in subsequent reads, providing a consistent view of the reader's content.
+	if br.index < int64(br.buffer.Len()) {
+		n := copy(out, br.buffer.Bytes()[br.index:])
+		br.index += int64(n)
+		return n, nil
+	}
+
+	if !br.activeBuffering {
+		// If buffering is not active, read directly from the underlying reader.
+		n, err := br.reader.Read(out)
+		br.index += int64(n)
+		br.bytesRead = max(br.bytesRead, br.index)
+		return n, err
+	}
+
+	// Ensure there are enough bytes in the buffer to read from.
+	if outLen+br.index > int64(br.buffer.Len()) {
+		bytesToRead := int(outLen + br.index - int64(br.buffer.Len()))
+		readerBytes := make([]byte, bytesToRead)
+		n, err := br.reader.Read(readerBytes)
+		br.buffer.Write(readerBytes[:n])
+		br.bytesRead += int64(n)
+
+		if err != nil {
+			return n, err
+		}
+	}
+
+	// Ensure the read does not exceed the buffer length.
+	endIndex := br.index + outLen
+	bufLen := int64(br.buffer.Len())
+	if endIndex > bufLen {
+		endIndex = bufLen
+	}
+
+	if br.index >= bufLen {
+		return 0, io.EOF
+	}
+
+	n := copy(out, br.buffer.Bytes()[br.index:endIndex])
+	br.index += int64(n)
+	return n, nil
+}
+
+// Seek sets the offset for the next Read or Write to offset.
+// It supports both seekable and non-seekable underlying readers.
+func (br *BufferedReadSeeker) Seek(offset int64, whence int) (int64, error) {
+	if br.seeker != nil {
+		// Use the underlying Seeker if available.
+		newIndex, err := br.seeker.Seek(offset, whence)
+		if err != nil {
+			return 0, fmt.Errorf("error seeking in reader: %w", err)
+		}
+		br.index = newIndex
+		if newIndex > br.bytesRead {
+			br.bytesRead = newIndex
+		}
+		return newIndex, nil
+	}
```

Comment:
Why do we need to track the index internally if the underlying reader is `io.Seeker`?

Reply:
We probably don't need it. I likely added it, thinking we needed it for a different reason, but I can't remember what that was.